### PR TITLE
Extract Source interface from SourceLookup

### PR DESCRIPTION
**Pull request by romseygeek**
SourceLookup combines a mutable lookup object that can be advanced
to different documents with access to a document's source.  This combination
can make reasoning about where a Source comes from difficult, particularly
in the FetchPhase where the source gets passed around a great deal.

This commit extracts a Source interface from SourceLookup, giving read-only
access to the source, and changes various FetchPhase interfaces to take this
read-only view instead of a full lookup.  You can now tell easily if a consumer 
of the source is going to try and move it to a different document.  As part of this 
change we add a new `docId` parameter to various ValueFetcher methods, as
previously this could be accessed via the SourceLookup.

### CODE REVIEW

**Review by nik9000 — APPROVED** [review]
I like it. Probably worth other folks approving as well.

File: server/src/main/java/org/elasticsearch/search/lookup/Source.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.search.fetch.subphase.FetchSourceContext;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.util.Map;
+
+/**
+ * The source of a document
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/search/lookup/Source.java:19]: The semantics here are a bit odd, in that if you call either `source()` or `internalSourceRef()` then the object you get returned is immutable and won't be swapped out from under you, but as things stand if the underlying reference is a SourceLookup then things may change before you call those methods.  I'll add some javadoc.  But yes, ideally `SourceLookup` will stop implementing this directly and then everything will be lovely and immutable.

**Review by jtibshirani — COMMENTED** [review]
I definitely see the motivation for this -- it's nice to enforce that the value fetchers won't modify the docId at all.

There's always a trade-off between adding another abstraction like this, and the benefits we get. As Nik pointed out, `Source` is a little confusing because it really looks immutable, but it's actually mutable. Do you have a follow-up plan to use it more broadly, or to pull it apart more from `SourceLookup`? Maybe you could share a rough idea.

**Comment by romseygeek** [comment]
Yes, there are several parts to this.  This first step is to re-assure callers of the various fetch methods that those callers won't then advance the SourceLookup out from underneath them.  Then we add some concrete implementations of `Source` that are just static wrappers around BytesReferences or Maps, which we can swap out for SourceLookup in a number of places.  Lastly we introduce a SourceProvider interface that returns a `Source` object given a reader context and doc id - possibly a global one as we have now, but I'd like to try a reader-specific LeafSourceLookup so that we can get some thread-safety.  The various implementations here will always return immutable Sources.  SourceLookup stops implementing `Source` and becomes an implementation detail of SearchLookup used only for scripts.

**Comment by romseygeek** [comment]
Hi @jtibshirani this PR has to temporarily revert https://github.com/elastic/elasticsearch/pull/90784 but my plan is to follow up by making HitContext take a Source as a constructor parameter rather than having a SourceLookup, which will essentially do the same thing.

**Comment by romseygeek** [comment]
@elasticmachine update branch

**Comment by jtibshirani** [comment]
@romseygeek sorry for the conflict, thought I had just found a low-hanging refactor while reviewing your PR! The plan makes a lot of sense.

**Review by jtibshirani — APPROVED** [review]
@romseygeek sorry for the conflict, thought I had just found a low-hanging refactor while reviewing your PR! The plan makes a lot of sense.

File: server/src/main/java/org/elasticsearch/index/mapper/ValueFetcher.java [hunk 2]
```diff
@@ -35,24 +35,26 @@ public interface ValueFetcher {
...
      * should not be relied on.
      *
-     * @param lookup        a lookup structure over the document's source.
+     * @param source        the document's source.
+     * @param doc           the document id
      * @param ignoredValues a mutable list to collect any ignored values as they were originally presented in source
      * @return a list a standardized field values.
      */
-    List<Object> fetchValues(SourceLookup lookup, List<Object> ignoredValues) throws IOException;
+    List<Object> fetchValues(Source source, int doc, List<Object> ignoredValues) throws IOException;
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/ValueFetcher.java:43]: This is a great step forward: I wonder long term whether we still want ValueFetcher to be source-centric. There are already implementations that don't retrieve from source, which with this change ignore the Source argument completely. I was planning on addressing some of this as a follow-up to #87267 but I never did.

**Comment by tylersmalley** [comment]
This change caused multiple e2e failures preventing the promotion of the ES snapshot. It's being tracked [here](https://github.com/elastic/kibana/issues/143339). 

Any help will be appreciated to understand if this is a regression or something that needs to be addressed on the Kibana side.

**Comment by romseygeek** [comment]
It should just be a refactoring that doesn't require any changes from Kibana.  Do you have stack traces from the errors you're seeing @tylersmalley?

**Comment by romseygeek** [comment]
Found it - we're passing the global docid to the doc value fetcher, not the reader-local docid.  And it passed CI because none of the tests index enough docs for us to have multiple leaves so the global docid and local docid were the same.

**Comment by romseygeek** [comment]
I opened https://github.com/elastic/elasticsearch/pull/90901

**Pull request merged**